### PR TITLE
Refactor GPUBindGroupLayout to isolate binding type definitions

### LINKED ISSUE #1164: There are too many combinations of texture bind point type and sampler bind point type
From https://github.com/gpuweb/gpuweb/pull/1076

We now have 5 texture bind point type and 3 sampler bind point type. Now that the matrix has grown so much, the claim "people just know what bind point to pick" is more difficult too justify. In particular, our matrix is now significantly larger than it is for any native API.

We should investigate just how costly deferring certain kinds of validation is.

I think there could be three different results from this analysis:
1. Deferred validation is too expensive, we shouldn't allow authors to trigger it, no way no how.
2. Deferred validation can get expensive, but for simple cases it's not, and the benefit of allowing authors to specify less information up front is sometimes larger than the runtime cost of deferred validation, for some apps
3. Deferred validation is somehow super cheap, and we can just always defer it and not require any of this complicated stuff up front

### PR DESCRIPTION
**Pull request by toji**
Fixes #1164 

This is currently a WIP because this change touches a very large number of references in the spec and I haven't yet updated them all. Wanted to get this pushed to the server for the weekend, though, and figured you might be interested in taking a peek. Of special interest is the updated binding type table and the bind group validation logic.

One thing to note early on: The usage of the term "kind" has felt rather awkward as I've been writing out some of the spec prose here. I don't have a suggested alternative, just point out that I've found it hard to write about naturally.

***
<a href="https://pr-preview.s3.amazonaws.com/toji/gpuweb/pull/1223.html" title="Last updated on Nov 24, 2020, 10:15 PM UTC (4cbd93e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1223/78e7c93...toji:4cbd93e.html" title="Last updated on Nov 24, 2020, 10:15 PM UTC (4cbd93e)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -2473,30 +2473,35 @@ dictionary GPUBindGroupLayoutDescriptor : GPUObjectDescriptorBase {
...
     required GPUIndex32 binding;
     required GPUShaderStageFlags visibility;
-    required GPUBindingType type;
-
-    // Used for uniform buffer and storage buffer bindings. Must be undefined for other binding types.
-    boolean hasDynamicOffset;
 
-    // Used for uniform buffer and storage buffer bindings. Must be undefined for other binding types.
-    GPUSize64 minBufferBindingSize;
+    // Used for uniform buffer and storage buffer bindings. Must be null for other binding types.
```
Comment (kainino0x) [inline spec/index.bs:2487]: Fields are always optional unless they're `required` (or have a default, `= {}`). We don't need to make these `null`able since the fields can be unset or `undefined`. (e.g. `{ binding: 0, visibility: 0, buffer: {} }` would be valid).

File: spec/index.bs [hunk 2]
```diff
@@ -2510,146 +2515,215 @@ Issue(https://github.com/gpuweb/gpuweb/issues/851): consider making `textureComp
...
+
+<script type=idl>
+enum GPUStorageTextureKind {
+    "readonly",
+    "writeonly",
+};
+
+dictionary GPUStorageTextureLayoutEntry {
+    GPUStorageTextureAccess kind;
+    GPUTextureFormat format;
```
Comment (kainino0x) [inline spec/index.bs:2700]: These two should be `required`

File: spec/index.bs (hunk 1 above)
Comment (kainino0x) [inline spec/index.bs:2487]: We should take this opportunity to move this stuff from comments to real prose. The reason it was comments was because this dictionary was so difficult to understand in the first place (understanding which fields were used for which binding types). Now that we're solving that problem, we can simply say just below here, "Exactly one of .buffer, .sampler, .texture, or .storageTexture must be specified. Which is specified, and value of that sub-entry, determines what type of resource can be bound. (Note: a value of `{}` is valid if the type has no required members.)`

**Review by kainino0x — COMMENTED** [review]
I only skimmed this so far, just a few IDL nits. But overall I think the API structure is great.

re: kind. We could just use `type`, since we are removing the other field called `type`. Semantically it would specify which "type" among the "types" allowed by the GPU*LayoutEntry type. (Maybe we need a good word for "(GPUBufferLayoutEntry|GPUSamplerLayoutEntry|GPUTextureLayoutEntry|GPUStorageTextureLayoutEntry)" though.)

**Comment by Kangz** [comment]
I was surprised to see `GPUStorageTextureLayoutEntry` split from `GPUTextureLayoutEntry`. Maybe they could stay the same?

Otherwise the "kind" naming could be replaced by "type" so that would be `GPUBufferBindingType` `GPUSamplerBindingType` and `GPUTextureBindingType`. Kind vs. type was useful in prior iterations of the rework but don't make sense now that things are in sub-descriptors. WDYT?

**Comment by kainino0x** [comment]
I suggested that they be separate. There's only some overlap between the options for sampled vs storage textures as they're quite different semantically. They both have `viewDimension`, but `samplingResultType` and `format` are unique to one or the other. Since sampled textures are much more common than storage textures, I think it's worth keeping the additional field and "kind" values separate.

**Comment by kainino0x** [comment]
Editors meeting: Agreed on new wording (samplers are "filtering" or "non-filtering", textures are "filterable" or "non-filterable"). Texture bindings are either "texture" (not "sampled texture" as multisampled textures can't be sampled) or "storage texture". Not quite decided on how multisampling works yet. Filing an issue about that.

**Comment by kainino0x** [comment]
And we also decided to keep `GPUStorageTextureLayoutEntry` split from `GPUTextureLayoutEntry`.

**Comment by toji** [comment]
Updated to address the initial feedback, use the new naming that @kainino0x mentioned, and fix and remaining broken references. Verbiage is generally better now but still a bit inconsistent in places. Let me know if anything sounds odd or unwieldy to you!

**Comment by Kangz** [comment]
I'm still a bit lukewarm about `GPUStorageTextureLayoutEntry` being separate but I think either way is fine. `unfilterable` should probably `non-filterable` so that it matches `non-filtering` but I understand I might have a higher tolerance to broken english than others :P

`GPUTextureComponentType` should be renamed because it doesn't really explain what `depth-comparison` it is. Instead we could rename it to `GPUSamplingOperationType` or `GPUSamplingOperationResult` (the result is a depth-comparison or the result if a float value)

**Comment by toji** [comment]
> `unfilterable` should probably `non-filterable` so that it matches `non-filtering` but I understand I might have a higher tolerance to broken english than others :P

`unfilterable` sounds ever-so-slightly more natural to me than `non-filterable`, but both sound a bit awkward and neither is so outright wrong that I'd balk at it. #NamingIsHard

**Comment by kainino0x** [comment]
> `unfilterable` sounds ever-so-slightly more natural to me than `non-filterable`, but both sound a bit awkward and neither is so outright wrong that I'd balk at it. #NamingIsHard

Same, both are ok with me.

**Comment by kainino0x** [comment]
> `GPUTextureComponentType` should be renamed because it doesn't really explain what `depth-comparison` it is. Instead we could rename it to `GPUSamplingOperationType` or `GPUSamplingOperationResult` (the result is a depth-comparison or the result if a float value)

We were going to rename it to `depth-stencil` because it's not comparison-only; you should also be able to read from depth/stencil. (I can't remember where the conversation with @kvark about this was, maybe only in meetings.) Then calling it component type makes sense.

File: spec/index.bs [hunk 3]
```diff
@@ -2510,146 +2504,227 @@ Issue(https://github.com/gpuweb/gpuweb/issues/851): consider making `textureComp
...
+    : <dfn>type</dfn>
+    ::
+        Indicates the required type of a sampler bound to this bindings.
+</dl>
+
+<script type=idl>
+enum GPUTextureType {
+    "filterable",
+    "unfilterable",
+    "multisample",
```
Comment (kainino0x) [inline spec/index.bs:2655]: idea: since they others describe what can be done ("-able") maybe this should be the same? e.g. "sample-addressable" or "multisample-addressable"

**Comment by kvark** [comment]
@kainino0x that was in https://github.com/gpuweb/gpuweb/issues/1198#issuecomment-724322605

**Comment by kainino0x** [comment]
I filed #1230. We can fix it in or after this PR, either way.

File: spec/index.bs [hunk 4]
```diff
@@ -2510,146 +2504,227 @@ Issue(https://github.com/gpuweb/gpuweb/issues/851): consider making `textureComp
...
+<dl dfn-type=dict-member dfn-for=GPUSamplerLayoutEntry>
+    : <dfn>type</dfn>
+    ::
+        Indicates the required type of a sampler bound to this bindings.
+</dl>
+
+<script type=idl>
+enum GPUTextureType {
+    "filterable",
+    "unfilterable",
```
Comment (kvark) [inline spec/index.bs:2654]: I think there is a problem with this approach. When using non-float textures, it would be nice if the users could rely on the default `GPUTextureType` and not specify it, because clearly it's redundant. Non-float textures can't be sampled in WGSL, so they naturally aren't filterable (and this term shouldn't even apply...). So we can't default this to `filterable`.

How about this strawman:
```webidl
enum GPUTextureType {
  "float",
  "unfilterable-float",
  "depth-float",
  "sint",
  "uint",
};
dictionary GPUTextureLayoutEntry {
    GPUTextureType type = "float";
    GPUTextureViewDimension viewDimension = "2d";
    bool multisampled = false;
};
```

Names are subject to change, of course :) The benefit of this is no need to specify redundant info for integer textures.

File: spec/index.bs [hunk 5]
```diff
@@ -2510,146 +2504,227 @@ Issue(https://github.com/gpuweb/gpuweb/issues/851): consider making `textureComp
...
+        Indicates the required {{GPUTextureViewDescriptor/dimension}} for texture views bound to
+        this binding.
+
+        Note:
+            This enables Metal-based WebGPU implementations to back the respective bind groups with
+            `MTLArgumentBuffer` objects that are more efficient to bind at run-time.
+</dl>
+
+<script type=idl>
+enum GPUStorageTextureType {
```
Comment (kvark) [inline spec/index.bs:2691]: nit: `GPUStorageTextureAccess`?

**Review by kvark — COMMENTED** [review]
Amazing work here, thank you @toji for iterating on this!

File: spec/index.bs [hunk 6]
```diff
@@ -2473,29 +2471,25 @@ dictionary GPUBindGroupLayoutDescriptor : GPUObjectDescriptorBase {
...
 
-    // Used for sampled texture and storage texture bindings. Must be undefined for other binding types.
-    GPUTextureViewDimension viewDimension;
-
-    // Used for sampled texture bindings. Must be undefined for other binding types.
-    GPUTextureComponentType textureComponentType;
-
-    // Used for storage texture bindings. Must be undefined for other binding types.
-    GPUTextureFormat storageTextureFormat;
+    GPUBufferLayoutEntry buffer;
```
[the remark this answers is not in the document]
Comment (toji) [reply spec/index.bs]: Sounds good. Changed the name of all four resource layout types, since the same principle seemed to apply to all of them.

File: spec/index.bs (hunk 4 above)
Comment (toji) [reply spec/index.bs:2654]: Shouldn't be a problem, but I'll leave it to others a bit better versed in the underlying APIs to greenlight this before I make the change.

**Comment by kainino0x** [comment]
I think this is on the main meeting agenda, but speculatively adding to the editors meeting agenda too.

File: spec/index.bs [hunk 7]
```diff
@@ -2228,7 +2226,7 @@ enum GPUTextureComponentType {
     "sint",
     "uint",
     // Texture is used with comparison sampling only.
-    "depth-comparison"
+    "depth-stencil"
```
[the remark this answers is not in the document]
Comment (Kangz) [reply spec/index.bs]: More specifically in https://github.com/gpuweb/gpuweb/issues/1230

File: spec/index.bs (hunk 4 above)
Comment (Kangz) [reply spec/index.bs:2654]: +1 to @kvark's suggestion. I think we recently (re-)realized that D3D12 prevented any form of sampling of integer textures so the filterable / unfilterable distinction is only for float textures.

The names `depth-float` and `GPUTextureType` seem a bit generic though, so I'd be open to bikeshed them. (though that could be in a follow-up if this is forcing too many iterations)

**Review by kvark — APPROVED** [review]
Thank you for updating!
The last concern about `GPUTextureType` still stands (we'll discuss it), but everything else looks great.

**Comment by toji** [comment]
Updated to apply @kvark's [suggestion](https://github.com/gpuweb/gpuweb/pull/1223#discussion_r527949749) after the structure of it was agreed on in the call (% a bit more name bikeshedding.)

**Review by kvark — CHANGES_REQUESTED** [review]
`GPUTextureComponentType` should now be removed

File: spec/index.bs [hunk 8]
```diff
@@ -2655,16 +2655,16 @@ dictionary GPUSamplerBindingLayout {
...
   "unfilterable-float",
-  "depth-float",
+  "depth",
   "sint",
   "uint",
 };
 
 dictionary GPUTextureBindingLayout {
-    GPUTextureType type = "float";
+    GPUTextureSampleType type = "float";
```
Comment (kainino0x) [inline spec/index.bs:2656]: Should we call this field `sampleType` or something?

File: spec/index.bs [hunk 9]
```diff
@@ -2808,7 +2808,7 @@ A {{GPUBindGroupLayout}} object has the following internal slots:
                                     - |textureLayout|.{{GPUTextureBindingLayout/viewDimension}} is
                                         {{GPUTextureViewDimension/"2d"}}.
                                     - |textureLayout|.{{GPUTextureBindingLayout/type}} is not
-                                        {{GPUTextureType/"depth-float"}}.
+                                        {{GPUTextureSampleType/"depth"}} or {{GPUTextureSampleType/"float"}}.
```
Comment (kainino0x) [inline spec/index.bs:2800]: I'm just (re-)noticing this particular rule. Shouldn't it be possible to load samples out of a multisampled depth texture? The rule may have been added when we thought "depth-comparison" bindings could be used only for comparison and not for loading.

This can be a followup but let's make sure to file an issue if we land this.

**Review by kainino0x — APPROVED** [review]

File: spec/index.bs (hunk 8 above)
Comment (toji) [reply spec/index.bs:2656]: That would break the pattern of every other binding layout member having a `type`, which is a nice bit of consistency. FWIW We currently have the `type` member of `GPUStorageTextureBindingLayout` be a `GPUStorageTextureAccess` enum, so it's already not a 1:1 enum/member name mapping.

File: spec/index.bs (hunk 9 above)
Comment (toji) [reply spec/index.bs:2800]: I wondered about that too after today's call. I'm happy to remove the `depth` case here.

File: spec/index.bs [hunk 10]
```diff
@@ -2685,68 +2756,54 @@ A {{GPUBindGroupLayout}} object has the following internal slots:
...
-                                        be a format which can support storage usage.
+                                    - The [$layout entry binding type$] of |bindingDescriptor| is not
+                                        {{GPUBufferType/"readonly-storage"}} or {{GPUStorageTextureAccess/"writeonly"}}.
+
+                                - If the |textureLayout| is not `undefined` and
+                                    |textureLayout|.{{GPUTextureBindingLayout/multisampled}} is `true`:
+                                    - |textureLayout|.{{GPUTextureBindingLayout/viewDimension}} is
+                                        {{GPUTextureViewDimension/"2d"}}.
+                                    - |textureLayout|.{{GPUTextureBindingLayout/type}} is not
+                                        {{GPUTextureSampleType/"float"}}.
```
Comment (kainino0x) [inline spec/index.bs:2800]: Rereading this after removing "depth" from this (sorry I didn't realize this earlier), I think this refactor had more consequences than we realized. Shouldn't it be totally fine to create a multisampled layout entry with any "float" as well? Technically anything you could do with `multisampled:true,type:"float"` you could do with `multisampled:true,type:"unfilterable-float"` but I don't think that means we have to enforce that.

OTOH `multisampled:true,type:"float"` would be extra restrictive for no reason (wouldn't allow unfilterable float formats) so maybe it's best not to allow it.

I'm trying to make a table to understand this...

**Comment by kainino0x** [comment]
OK, this is excessively detailed, but here's a spreadsheet of the possible classes of textures and how you would use them in a layout.
https://docs.google.com/spreadsheets/d/12u6g6-dUGxuwgaNVRDdZVZ0u2zb9kotqEO3_ZA-1930/edit?usp=sharing

Note there's a new, smaller problem similar to the one @kvark identified (https://github.com/gpuweb/gpuweb/pull/1223#discussion_r527949749) which is that if you set multisampled:true, then leaving the GPUTextureSampleType as default ("float") is invalid (`multisampled:true,type:"float"`). We could make it valid if we want, though, and it would have exactly the same meaning as `multisampled:true,type:"unfilterable-float"`.

**Comment by kainino0x** [comment]
tl;dr I think what we have now is fine, but maybe consider making (true,"float) valid and definitely consider cases we want to support in the future

**Review by kvark — COMMENTED** [review]
I think the changes to the table are not complete. First, there is still this sentence to be removed:
> The "Filter" column specifies whether textures of this format can be sampled in shaders with a sampler having ...

Secondly, for filterable formats we want to list both "float" and "unfilterable-float" as supported sample types.

**Comment by toji** [comment]
> I think the changes to the table are not complete...

Done!

**Review by kvark — APPROVED** [review]
Thank you!
